File: net/JetBrains.Serialization/src/FileInfos/MachO/MachoFileInfo.cs
using JetBrains.FormatRipper.MachO;
using JetBrains.SignatureVerifier.Crypt;
using Newtonsoft.Json;

namespace JetBrains.Serialization.FileInfos.MachO;

[JsonObject(MemberSerialization.OptIn)]
public class MachoFileInfo : FileInfo
{
  [JsonProperty("FileMetaInfo")] public override IFileMetaInfo FileMetaInfo { get; }
  [JsonProperty("SignedDataInfo")] public override SignedDataInfo SignedDataInfo { get; }

  public MachoFileInfo(MachOFile.Section section)
  {
    var signedMessage = SignedMessage.CreateInstance(section.SignatureData);
    SignedDataInfo = new SignedDataInfo(signedMessage.SignedData);
    if (section.Metadata == null)
      throw new Exception("Metadata can not be null");

    FileMetaInfo = new MachoFileMetaInfo(section.Metadata);
  }

  [JsonConstructor]
  public MachoFileInfo(IFileMetaInfo fileMetaInfo, SignedDataInfo signedDataInfo)
  {
    FileMetaInfo = fileMetaInfo;
    SignedDataInfo = signedDataInfo;
  }

  public override void ModifyFile(Stream stream)
  {
    FileMetaInfo.ModifyFile(stream, SignedDataInfo.ToSignature("BER"));
  }
}

File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/fileInfos/MachO/MachoFileInfo.kt
package com.jetbrains.signatureverifier.serialization.fileInfos

import com.jetbrains.signatureverifier.crypt.SignedMessage
import com.jetbrains.signatureverifier.macho.MachoFile
import kotlinx.serialization.Serializable
import java.nio.channels.SeekableByteChannel

@Serializable
class MachoFileInfo : FileInfo {
  override val fileMetaInfo: FileMetaInfo
  override val signedDataInfo: SignedDataInfo

  constructor(machoFile: MachoFile) {
    val signatureData = machoFile.GetSignatureData()
    val signedMessage = SignedMessage.CreateInstance(signatureData)
    val signedData = signedMessage.SignedData

    signedDataInfo = SignedDataInfo(signedData)
    fileMetaInfo = machoFile.metaInfo
  }

  override fun modifyFile(stream: SeekableByteChannel) {
    fileMetaInfo.modifyFile(stream, signedDataInfo.toSignature("BER"))
  }
}